Improvements to Java API handling of datatypes and arrays

andrewcmyers opened the pull request:
This is a set of connected changes broken out from #11535 (Generic Java API) and improved
a bit further; still synergistic with that PR  and should allow whittling that PR down toward a digestible size.

* Changed how datatypes are computed from arrays to not crawl over the array but instead just inspect the array's class object. This seems somewhat simpler, more efficient, and easier to maintain going forward.
* Added the notion that a given array component type can correspond to more than one tensor datatype. I think that we already would like this to be able to support both uint8 and string using byte arrays, and the need will become more acute as more TensorFlow types are added.
* Resurrected the test case for creating and extracting uint8 tensors (requires previous change)

andrewcmyers committed:
This is a set of changes broken out from #11535 (Generic Java API) and improved
a bit further; still synergistic with that PR.

* Changed how datatypes are computed to crawl over the array.

* Added the notion that a given array component type can correspond to more than one
  tensor datatype. We already would like this to be able to support both uint8 and string
  using byte arrays; the need will become more acute as more TensorFlow types are added.

* Resurrected the test case for creating and extracting uint8 tensors.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

andrewcmyers commented:
The resurrected test case was written for #12797 but wouldn't work without this additional flexibility for tensor creation.

andrewcmyers commented on tensorflow/java/src/main/java/org/tensorflow/Tensor.java:
```diff
@@ -537,56 +548,72 @@ private static void throwExceptionIfNotByteOfByteArrays(Object array) {
+    if (!c.isArray()) return 0;
+    String s = c.getName();
+    int i = 0;
+    while (s.charAt(i) == '[') {
```
This could also be done by walking up the `getComponentType()` path but I'm conjecturing it's a little faster to just walk through the string.

It should also be correct to initialize i to 1 instead of 0 since the first character is guaranteed to be `[`, but starting from zero seemed more obviously correct .

asimshankar replied:
I don't have a strong opinion either way (though is there any reason to believe that one would be significantly faster than the other?)

asimshankar commented on tensorflow/java/src/main/java/org/tensorflow/Tensor.java:
```diff
@@ -537,56 +548,72 @@ private static void throwExceptionIfNotByteOfByteArrays(Object array) {
-      }
-      return 1 + numDimensions(e);
+  /** Returns the number of dimensions of a tensor of type dtype
+   *  when represented by the object o.
```
Mind running the [source code formatter](https://github.com/google/google-java-format) on the files?

andrewcmyers replied:
Done.

asimshankar reviewed (changes requested):
Thanks for splitting this out, looks great, just a few minor nits.

andrewcmyers committed:
Addressed Asim's comments (thanks!).

- Renamed class_datatypes and made it private.
- Changed dataTypeOf to walk the class objects instead of the class name. Seems clearer.
- Ran the code formatter.

andrewcmyers commented on tensorflow/java/src/main/java/org/tensorflow/Tensor.java:
```diff
@@ -537,56 +548,72 @@ private static void throwExceptionIfNotByteOfByteArrays(Object array) {
     }
   }
 
+  static HashMap<Class<?>, DataType> class_datatypes = new HashMap<>();
```
Done.

andrewcmyers commented on tensorflow/java/src/main/java/org/tensorflow/Tensor.java:
```diff
@@ -537,56 +548,72 @@ private static void throwExceptionIfNotByteOfByteArrays(Object array) {
     }
   }
 
+  static HashMap<Class<?>, DataType> class_datatypes = new HashMap<>();
```
Done.

asimshankar reviewed (approved):
Thanks!

asimshankar commented:
Jenkins, test this please

drpngx commented:
Windows should be fixed now.

Jenkins, test this please.
